File: InterviewQuestions/DS/tree.js
class Node {
    constructor(val) {
        this.value = val;
        this.left = null;
        this.right = null;
    }
}

class BST {
    constructor() {
        this.root = null;
    }

    // Inserts a new unique value in the BST
    insert(val) {
        const newNode = new Node(val);
        if (this.root == null) {
            this.root = newNode;
        }
        else {
            let root = this.root;
            let prev = null;
            while (root != null) {      //find correct position for insertion
                if (root.value === val) return undefined; // Handle Duplicates
                prev = root;
                root = val > root.value ? root.right : root.left;
            }
            if (val > prev.value)
                prev.right = newNode;
            else
                prev.left = newNode;
        }
        return this;
    }

    //find if the value is present in the BST
    find(val) {
        if (this.root != null) {
            let current = this.root;
            while (current != null) {
                if (current.value === val) return current;
                current = val > current.value ? current.right : current.left;
            }
        }
        return false;
    }

    printLeftView(root){
        if(root == null) 
        return;
        console.log(root.value)
        if(root.left){
            this.printLeftView(root.left)
        }
        if(root.right){
            this.printLeftView(root.right.left)
        }
    }

    printRightView(root){
        if(root == null) 
        return;
        console.log(root.value)
        if(root.right){
            this.printRightView(root.right)
        }
        if(root.left){
            this.printRightView(root.left.right)
        }
    }
}



let bst = new BST()

bst.insert(10)
bst.insert(2)
bst.insert(3)
bst.insert(7)
bst.insert(8)
bst.insert(15)
bst.insert(12)
bst.insert(14)

//bst.printLeftView(bst.root);
bst.printRightView(bst.root);

File: JavaInterviewQuestions/Tree.java
package helloworld;

class Node{
	int value;
	Node right;
	Node left;
	
	Node(int data){
		this.value = data;
		this.right = null;
		this.left = null;
	}
}

public class Tree {
	
	Node root;
	Tree(){
		this.root = null;
	}
	
	void insert(int data) {
		Node n = new Node(data);
		if(root == null)
			root = n;
		else {
			Node cur = root;
			Node prev = null;
			while(cur != null) {
				prev = cur;
				cur = data > cur.value ? cur.right : cur.left;
			}
			if(data > prev.value)
				prev.right = n;
			else
				prev.left = n;
		}
	}
	
	void printTree(Node n) {
		if(n == null)
			return;
		printTree(n.right);
		System.out.println(n.value);
		printTree(n.left);
	}
	
	void preOrder(Node n) {
		System.out.print(n.value + " ");
		if(n.left != null)
			preOrder(n.left);
		if(n.right != null)
			preOrder(n.right);
	}
	
	void inOrder(Node n) {
		if(n.left != null)
			inOrder(n.left);
		System.out.print(n.value + " ");
		if(n.right != null)
			inOrder(n.right);
	}
	
	void postOrder(Node n) {
		if(n.left != null)
			preOrder(n.left);
		if(n.right != null)
			preOrder(n.right);
		System.out.print(n.value + " ");
	}
	
	public static void main(String args[]) {
		Tree ob = new Tree();
		ob.insert(5);
		ob.insert(4);
		ob.insert(7);
		ob.insert(2);
		ob.insert(8);
		//ob.printTree(ob.root);
		ob.preOrder(ob.root);
		System.out.println(" ");
		ob.inOrder(ob.root);
		System.out.println(" ");
		ob.postOrder(ob.root);
	}
	
}
